in_toto/envelope.go

```diff
@@ -0,0 +1,197 @@
+package in_toto
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/secure-systems-lab/go-securesystemslib/cjson"
+	"github.com/secure-systems-lab/go-securesystemslib/dsse"
+	"github.com/secure-systems-lab/go-securesystemslib/signerverifier"
+)
+
+// PayloadType is the payload type used for links and layouts.
+const PayloadType = "application/vnd.in-toto+json"
+
+// ErrInvalidPayloadType indicates that the envelope used an unknown payload type
+var ErrInvalidPayloadType = errors.New("unknown payload type")
+
+type Envelope struct {
+	envelope          *dsse.Envelope
+	payload           any
+	decodePayloadOnce sync.Once
+}
+
+func NewEnvelope() *Envelope {
+	return &Envelope{envelope: &dsse.Envelope{}}
+}
+
+func (e *Envelope) SetPayload(payload any) error {
+	encodedBytes, err := cjson.EncodeCanonical(payload)
+	if err != nil {
+		return err
+	}
+
+	e.payload = payload
+	e.envelope = &dsse.Envelope{
+		Payload:     base64.StdEncoding.EncodeToString(encodedBytes),
+		PayloadType: PayloadType,
+	}
+
+	return nil
+}
+
+func (e *Envelope) GetPayload() any {
+	e.decodePayloadOnce.Do(func() {
+		payloadBytes, err := e.envelope.DecodeB64Payload()
+		if err != nil {
+			panic(fmt.Errorf("error decoding payload: %w", err))
+		}
+
+		var payload map[string]any
+		if err := json.Unmarshal(payloadBytes, &payload); err != nil {
+			panic(fmt.Errorf("error decoding payload: %w", err))
+		}
+
+		if payload["_type"] == "link" {
+			var link Link
+			if err := checkRequiredJSONFields(payload, reflect.TypeOf(link)); err != nil {
+				panic(fmt.Errorf("error decoding payload: %w", err))
```

Comment:
could you return the error and handle it upstream instead? Having a Panic here doesn't feel right.

Reply:
I'm not sure if `Do()` can pass that up, I agree about panic. Should we rework this to get away with sync.Once?

Reply:
hmmm....I think that would be cleaner. How much work would that take?

Reply:
let me see.

Reply:
https://github.com/in-toto/in-toto-golang/pull/228/commits/71d35ad4a95b009588d473f396bdf0bdc0a463b1 has a tentative patch for this.